{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import os\n",
    "\n",
    "import glob\n",
    "import sqlalchemy as sa\n",
    "from sqlalchemy.orm import Mapped, relationship, DeclarativeBase, mapped_column, Session\n",
    "\n",
    "from opengsync_db.models.Base import Base\n",
    "from opengsync_db import categories, DBHandler, models\n",
    "from opengsync_db import units\n",
    "\n",
    "from flask_bcrypt import Bcrypt\n",
    "\n",
    "from loguru import logger\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "import os\n",
    "load_dotenv()\n",
    "\n",
    "bcrypt = Bcrypt()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'lbi-dev.int.cemm.at'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.environ[\"POSTGRES_SERVER_IP\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOG: Connection closed.\n",
      "LOG: Connected to DB 'postgresql+psycopg://lbi-dev.int.cemm.at:5432/opengsync_db'\n"
     ]
    }
   ],
   "source": [
    "db = DBHandler(expire_on_commit=False)\n",
    "# db.connect(os.environ[\"POSTGRES_USER\"], os.environ[\"POSTGRES_PASSWORD\"], \"localhost\", port=5432)\n",
    "db.connect(\n",
    "    user=os.environ[\"POSTGRES_USER\"],\n",
    "    password=os.environ[\"POSTGRES_PASSWORD\"],\n",
    "    host=os.environ[\"POSTGRES_SERVER_IP\"],\n",
    "    port=os.environ[\"POSTGRES_PORT\"],\n",
    "    db=os.environ[\"POSTGRES_DB\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "label_search_columns: dict[str, list[str]] = {\n",
    "    str(models.Project.__tablename__): [\"identifier\", \"title\"],\n",
    "    str(models.SeqRequest.__tablename__): [\"name\"],\n",
    "    str(models.Library.__tablename__): [\"name\"],\n",
    "    str(models.Pool.__tablename__): [\"name\"],\n",
    "    str(models.Experiment.__tablename__): [\"name\"],\n",
    "    str(models.SeqRun.__tablename__): [\"experiment_name\"],\n",
    "    str(models.Kit.__tablename__): [\"name\", \"identifier\"],\n",
    "    str(models.User.__tablename__): [\"email\", \"last_name\", \"first_name\"],\n",
    "    str(models.Group.__tablename__): [\"name\"],\n",
    "    str(models.Feature.__tablename__): [\"name\", \"target_name\", \"target_id\"],\n",
    "    str(models.SeqRun.__tablename__): [\"experiment_name\"],\n",
    "    str(models.Plate.__tablename__): [\"name\"],\n",
    "    str(models.Barcode.__tablename__): [\"sequence\", \"name\"],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "project\n",
      " - id\n",
      " - identifier\n",
      " - title\n",
      " - description\n",
      " - timestamp_created_utc\n",
      " - status_id\n",
      " - num_samples\n",
      " - owner_id\n",
      " - group_id\n",
      " - software\n",
      "sample\n",
      " - id\n",
      " - name\n",
      " - status_id\n",
      " - num_libraries\n",
      " - qubit_concentration\n",
      " - avg_fragment_size\n",
      " - timestamp_stored_utc\n",
      " - project_id\n",
      " - ba_report_id\n",
      " - owner_id\n",
      " - attributes\n",
      "user_affiliation\n",
      " - user_id\n",
      " - group_id\n",
      " - affiliation_type_id\n",
      "sample_plate_link\n",
      " - plate_id\n",
      " - well_idx\n",
      " - sample_id\n",
      " - library_id\n",
      "sample_library_link\n",
      " - mux\n",
      " - sample_id\n",
      " - library_id\n",
      "lane_pool_link\n",
      " - lane_id\n",
      " - pool_id\n",
      " - experiment_id\n",
      " - dilution_id\n",
      " - num_m_reads\n",
      " - lane_num\n",
      "library_feature_link\n",
      " - library_id\n",
      " - feature_id\n",
      "seq_request_delivery_email_link\n",
      " - seq_request_id\n",
      " - email\n",
      " - status_id\n",
      "experiment\n",
      " - id\n",
      " - name\n",
      " - timestamp_created_utc\n",
      " - timestamp_finished_utc\n",
      " - r1_cycles\n",
      " - r2_cycles\n",
      " - i1_cycles\n",
      " - i2_cycles\n",
      " - workflow_id\n",
      " - status_id\n",
      " - operator_id\n",
      " - sequencer_id\n",
      "pool\n",
      " - id\n",
      " - name\n",
      " - status_id\n",
      " - type_id\n",
      " - timestamp_stored_utc\n",
      " - clone_number\n",
      " - original_pool_id\n",
      " - num_m_reads_requested\n",
      " - avg_fragment_size\n",
      " - qubit_concentration\n",
      " - num_libraries\n",
      " - owner_id\n",
      " - plate_id\n",
      " - seq_request_id\n",
      " - contact_id\n",
      " - ba_report_id\n",
      " - lab_prep_id\n",
      " - experiment_id\n",
      "lims_user\n",
      " - id\n",
      " - first_name\n",
      " - last_name\n",
      " - email\n",
      " - password\n",
      " - role_id\n",
      " - num_projects\n",
      " - num_samples\n",
      " - num_seq_requests\n",
      "seq_request\n",
      " - id\n",
      " - name\n",
      " - description\n",
      " - special_requirements\n",
      " - billing_code\n",
      " - data_delivery_mode_id\n",
      " - read_type_id\n",
      " - submission_type_id\n",
      " - status_id\n",
      " - timestamp_submitted_utc\n",
      " - timestamp_finished_utc\n",
      " - read_length\n",
      " - num_lanes\n",
      " - num_libraries\n",
      " - organization_contact_id\n",
      " - requestor_id\n",
      " - group_id\n",
      " - bioinformatician_contact_id\n",
      " - contact_person_id\n",
      " - billing_contact_id\n",
      " - sample_submission_event_id\n",
      "library\n",
      " - id\n",
      " - name\n",
      " - sample_name\n",
      " - clone_number\n",
      " - original_library_id\n",
      " - type_id\n",
      " - status_id\n",
      " - genome_ref_id\n",
      " - assay_type_id\n",
      " - mux_type_id\n",
      " - index_type_id\n",
      " - timestamp_stored_utc\n",
      " - nuclei_isolation\n",
      " - seq_depth_requested\n",
      " - avg_fragment_size\n",
      " - volume\n",
      " - qubit_concentration\n",
      " - num_samples\n",
      " - num_features\n",
      " - properties\n",
      " - ba_report_id\n",
      " - pool_id\n",
      " - experiment_id\n",
      " - owner_id\n",
      " - seq_request_id\n",
      " - lab_prep_id\n",
      "kit\n",
      " - id\n",
      " - name\n",
      " - identifier\n",
      " - kit_type_id\n",
      "index_kit\n",
      " - id\n",
      " - type_id\n",
      " - supported_protocol_ids\n",
      "contact\n",
      " - id\n",
      " - name\n",
      " - email\n",
      " - phone\n",
      " - address\n",
      "sequencer\n",
      " - id\n",
      " - name\n",
      " - model_id\n",
      " - ip\n",
      "adapter\n",
      " - id\n",
      " - well\n",
      " - index_kit_id\n",
      "feature\n",
      " - id\n",
      " - name\n",
      " - sequence\n",
      " - pattern\n",
      " - read\n",
      " - target_name\n",
      " - target_id\n",
      " - type_id\n",
      " - feature_kit_id\n",
      "feature_kit\n",
      " - id\n",
      " - type_id\n",
      "file\n",
      " - id\n",
      " - name\n",
      " - extension\n",
      " - type_id\n",
      " - uuid\n",
      " - size_bytes\n",
      " - timestamp_utc\n",
      " - uploader_id\n",
      " - seq_request_id\n",
      " - experiment_id\n",
      " - lab_prep_id\n",
      "seqquality\n",
      " - id\n",
      " - lane\n",
      " - num_lane_reads\n",
      " - num_library_reads\n",
      " - mean_quality_pf_r1\n",
      " - q30_perc_r1\n",
      " - mean_quality_pf_r2\n",
      " - q30_perc_r2\n",
      " - mean_quality_pf_i1\n",
      " - q30_perc_i1\n",
      " - mean_quality_pf_i2\n",
      " - q30_perc_i2\n",
      " - library_id\n",
      " - experiment_id\n",
      "comment\n",
      " - id\n",
      " - text\n",
      " - timestamp_utc\n",
      " - file_id\n",
      " - author_id\n",
      " - seq_request_id\n",
      " - experiment_id\n",
      " - lab_prep_id\n",
      "seq_run\n",
      " - id\n",
      " - experiment_name\n",
      " - status_id\n",
      " - instrument_name\n",
      " - run_folder\n",
      " - flowcell_id\n",
      " - read_type_id\n",
      " - rta_version\n",
      " - recipe_version\n",
      " - side\n",
      " - flowcell_mode\n",
      " - r1_cycles\n",
      " - r2_cycles\n",
      " - i1_cycles\n",
      " - i2_cycles\n",
      " - quantities\n",
      "lane\n",
      " - id\n",
      " - number\n",
      " - phi_x\n",
      " - avg_fragment_size\n",
      " - original_qubit_concentration\n",
      " - sequencing_qubit_concentration\n",
      " - total_volume_ul\n",
      " - library_volume_ul\n",
      " - target_molarity\n",
      " - experiment_id\n",
      " - ba_report_id\n",
      "pool_dilution\n",
      " - id\n",
      " - identifier\n",
      " - pool_id\n",
      " - timestamp_utc\n",
      " - operator_id\n",
      " - qubit_concentration\n",
      " - volume_ul\n",
      "plate\n",
      " - id\n",
      " - name\n",
      " - num_cols\n",
      " - num_rows\n",
      " - owner_id\n",
      " - lab_prep_id\n",
      "barcode\n",
      " - id\n",
      " - sequence\n",
      " - name\n",
      " - well\n",
      " - type_id\n",
      " - adapter_id\n",
      " - index_kit_id\n",
      "library_index\n",
      " - id\n",
      " - library_id\n",
      " - index_kit_i7_id\n",
      " - index_kit_i5_id\n",
      " - name_i7\n",
      " - name_i5\n",
      " - sequence_i7\n",
      " - sequence_i5\n",
      "lab_prep\n",
      " - id\n",
      " - name\n",
      " - prep_number\n",
      " - protocol_id\n",
      " - status_id\n",
      " - assay_type_id\n",
      " - creator_id\n",
      "event\n",
      " - id\n",
      " - title\n",
      " - note\n",
      " - timestamp_utc\n",
      " - type_id\n",
      " - creator_id\n",
      "group\n",
      " - id\n",
      " - name\n",
      " - type_id\n"
     ]
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "SELECT * FROM pg_catalog.pg_tables;\n",
    "\"\"\"\n",
    "df = pd.read_sql(q, db._engine)\n",
    "\n",
    "os.makedirs(\"init\", exist_ok=True)\n",
    "\n",
    "with open(os.path.join(\"init\", \"tables.txt\"), \"w\") as f:\n",
    "    for table in Base.metadata.tables.items():\n",
    "        table_name = table[0]\n",
    "        f.write(f\"{table_name}\\n\")\n",
    "        if table_name not in df[\"tablename\"].values:\n",
    "            raise Exception(f\"Table {table[0]} is missing from the DB.\")\n",
    "\n",
    "        print(table_name)\n",
    "        for column in table[1].columns:\n",
    "            column_name = column.name\n",
    "            f.write(f\" - {column_name}\\n\")\n",
    "            print(f\" - {column_name}\")\n",
    "            q = f\"SELECT column_name FROM information_schema.columns WHERE table_name='{table_name}' and column_name='{column_name}';\"\n",
    "            if len(pd.read_sql(q, db._engine)) == 0:\n",
    "                raise Exception(f\"Column {column_name} is missing from the table {table_name}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking if pg_trgm extension is installed.\n",
      "pg_trgm extension is installed.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>oid</th>\n",
       "      <th>extname</th>\n",
       "      <th>extowner</th>\n",
       "      <th>extnamespace</th>\n",
       "      <th>extrelocatable</th>\n",
       "      <th>extversion</th>\n",
       "      <th>extconfig</th>\n",
       "      <th>extcondition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16981</td>\n",
       "      <td>pg_trgm</td>\n",
       "      <td>10</td>\n",
       "      <td>2200</td>\n",
       "      <td>True</td>\n",
       "      <td>1.6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     oid  extname  extowner  extnamespace  extrelocatable extversion  \\\n",
       "0  16981  pg_trgm        10          2200            True        1.6   \n",
       "\n",
       "  extconfig extcondition  \n",
       "0      None         None  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extensions\n",
    "q = \"\"\"\n",
    "SELECT * FROM pg_extension WHERE extname = 'pg_trgm';\n",
    "\"\"\"\n",
    "\n",
    "print(\"Checking if pg_trgm extension is installed.\")\n",
    "extensions = pd.read_sql(q, db._engine)\n",
    "if len(extensions) == 0:\n",
    "    print(\"Installing pg_trgm extension.\")\n",
    "\n",
    "    with db._engine.connect() as conn:\n",
    "        conn.execute(sa.text('CREATE EXTENSION pg_trgm;COMMIT;'))\n",
    "else:\n",
    "    print(\"pg_trgm extension is installed.\")\n",
    "\n",
    "pd.read_sql(q, db._engine)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Indexes\n",
    "\n",
    "# Postgres full text search\n",
    "for table, columns in label_search_columns.items():\n",
    "    for column in columns:\n",
    "        with db._engine.connect() as conn:\n",
    "            conn.execute(sa.text(f\"\"\"\n",
    "                CREATE INDEX IF NOT EXISTS\n",
    "                    trgm_{table}_{column}_idx\n",
    "                ON\n",
    "                    \"{table}\"\n",
    "                USING\n",
    "                    gin (lower({column}) gin_trgm_ops);COMMIT;\n",
    "            \"\"\"))\n",
    "\n",
    "with db._engine.connect() as conn:\n",
    "    conn.execute(sa.text(\"\"\"\n",
    "        CREATE INDEX IF NOT EXISTS\n",
    "            trgm_user_full_name_idx\n",
    "        ON\n",
    "            \"lims_user\"\n",
    "        USING\n",
    "            gin ((first_name || ' ' || last_name) gin_trgm_ops);COMMIT;\n",
    "    \"\"\"))\n",
    "\n",
    "    conn.execute(sa.text(\"\"\"\n",
    "        CREATE INDEX IF NOT EXISTS\n",
    "            trgm_kit_identifier_name_idx\n",
    "        ON\n",
    "            \"kit\"\n",
    "        USING\n",
    "            gin ((identifier || ' ' || name) gin_trgm_ops);COMMIT;\n",
    "    \"\"\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "opengsync",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
