{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO @ 2025-07-14 13:54:39 (__init__.py:78 in <module>):\n",
      ">   Sample submission window: Wednesday 15:00:00 - 17:00:00\n",
      "INFO @ 2025-07-14 13:54:39 (__init__.py:78 in <module>):\n",
      ">   Sample submission window: Friday 13:00:00 - 15:00:00\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'header': {'workflow': 'library_annotation',\n",
       "  'timestamp': '2025-07-14 13:40:16'},\n",
       " 'project_select': StepFile(args={'workflow_type': 'pooled'}, metadata={'project_name': 'Effect of a treatment on mouse CDX syngenic model', 'workflow': 'library_annotation', 'workflow_type': 'pooled', 'project_id': None, 'seq_request_id': 1, 'user_id': 3, 'project_description': 'Compared transcriptomic of treated CDX vs control\\n'}, tables={}),\n",
       " 'pooled_library_annotation': StepFile(args={}, metadata={'project_name': 'Effect of a treatment on mouse CDX syngenic model', 'workflow': 'library_annotation', 'workflow_type': 'pooled', 'project_id': None, 'seq_request_id': 1, 'user_id': 3, 'project_description': 'Compared transcriptomic of treated CDX vs control\\n', 'assay_type_id': 15, 'mux_type_id': None, 'nuclei_isolation': False}, tables={'sample_pooling_table':        sample_name                library_name mux_type_id\n",
       " 0  CDX_Melanoma_EC  CDX_Melanoma_EC_10XFLEXGEX        None, 'library_table':                  library_name      sample_name          genome  genome_id  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC  Mouse (GRCm38)          2   \n",
       " \n",
       "                library_type  library_type_id             pool seq_depth  \n",
       " 0  10X Flex Gene Expression                1  CDX_Melanoma_EC      None  , 'sample_table':        sample_name sample_id\n",
       " 0  CDX_Melanoma_EC      None}),\n",
       " 'pool_mapping': StepFile(args={}, metadata={'project_name': 'Effect of a treatment on mouse CDX syngenic model', 'workflow': 'library_annotation', 'workflow_type': 'pooled', 'project_id': None, 'seq_request_id': 1, 'user_id': 3, 'project_description': 'Compared transcriptomic of treated CDX vs control\\n', 'assay_type_id': 15, 'mux_type_id': None, 'nuclei_isolation': False, 'pool_contact_name': 'Emilie Crouchet', 'pool_contact_email': 'ecrouchet@unistra.fr', 'pool_contact_phone': '+33(0)368853707'}, tables={'sample_pooling_table':        sample_name                library_name mux_type_id\n",
       " 0  CDX_Melanoma_EC  CDX_Melanoma_EC_10XFLEXGEX        None, 'library_table':                  library_name      sample_name          genome  genome_id  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC  Mouse (GRCm38)          2   \n",
       " \n",
       "                library_type  library_type_id             pool seq_depth  \n",
       " 0  10X Flex Gene Expression                1  CDX_Melanoma_EC      None  , 'sample_table':        sample_name sample_id\n",
       " 0  CDX_Melanoma_EC      None, 'pool_table':          pool_name       pool_label pool_id  num_m_reads_requested\n",
       " 0  CDX_Melanoma_EC  CDX_Melanoma_EC    None                  400.0}),\n",
       " 'barcode_input': StepFile(args={}, metadata={'project_name': 'Effect of a treatment on mouse CDX syngenic model', 'workflow': 'library_annotation', 'workflow_type': 'pooled', 'project_id': None, 'seq_request_id': 1, 'user_id': 3, 'project_description': 'Compared transcriptomic of treated CDX vs control\\n', 'assay_type_id': 15, 'mux_type_id': None, 'nuclei_isolation': False, 'pool_contact_name': 'Emilie Crouchet', 'pool_contact_email': 'ecrouchet@unistra.fr', 'pool_contact_phone': '+33(0)368853707'}, tables={'sample_pooling_table':        sample_name                library_name mux_type_id\n",
       " 0  CDX_Melanoma_EC  CDX_Melanoma_EC_10XFLEXGEX        None, 'library_table':                  library_name      sample_name          genome  genome_id  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC  Mouse (GRCm38)          2   \n",
       " \n",
       "                library_type  library_type_id             pool seq_depth  \n",
       " 0  10X Flex Gene Expression                1  CDX_Melanoma_EC      None  , 'sample_table':        sample_name sample_id\n",
       " 0  CDX_Melanoma_EC      None, 'pool_table':          pool_name       pool_label pool_id  num_m_reads_requested\n",
       " 0  CDX_Melanoma_EC  CDX_Melanoma_EC    None                  400.0, 'barcode_table':   index_well   kit_i7 name_i7 sequence_i7   kit_i5 name_i5 sequence_i5  \\\n",
       " 0         A6  1000251    None        None  1000251    None        None   \n",
       " \n",
       "                  library_name kit_i7_id kit_i7_name kit_i5_id kit_i5_name  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX      None        None      None        None   \n",
       " \n",
       "   index_type_id  \n",
       " 0          None  }),\n",
       " 'index_kit_mapping': StepFile(args={}, metadata={'project_name': 'Effect of a treatment on mouse CDX syngenic model', 'workflow': 'library_annotation', 'workflow_type': 'pooled', 'project_id': None, 'seq_request_id': 1, 'user_id': 3, 'project_description': 'Compared transcriptomic of treated CDX vs control\\n', 'assay_type_id': 15, 'mux_type_id': None, 'nuclei_isolation': False, 'pool_contact_name': 'Emilie Crouchet', 'pool_contact_email': 'ecrouchet@unistra.fr', 'pool_contact_phone': '+33(0)368853707'}, tables={'sample_pooling_table':        sample_name                library_name mux_type_id\n",
       " 0  CDX_Melanoma_EC  CDX_Melanoma_EC_10XFLEXGEX        None, 'library_table':                  library_name      sample_name          genome  genome_id  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC  Mouse (GRCm38)          2   \n",
       " \n",
       "                library_type  library_type_id             pool seq_depth  \n",
       " 0  10X Flex Gene Expression                1  CDX_Melanoma_EC      None  , 'sample_table':        sample_name sample_id\n",
       " 0  CDX_Melanoma_EC      None, 'pool_table':          pool_name       pool_label pool_id  num_m_reads_requested\n",
       " 0  CDX_Melanoma_EC  CDX_Melanoma_EC    None                  400.0, 'barcode_table':                  library_name index_well   kit_i7   name_i7 sequence_i7  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX         A6  1000251  SI-TS-A6  AAAGTTCTGA   \n",
       " \n",
       "     kit_i5   name_i5 sequence_i5  kit_i7_id  kit_i5_id  \\\n",
       " 0  1000251  SI-TS-A6  CGCAGTCCAC         22         22   \n",
       " \n",
       "                             kit_i7_name                           kit_i5_name  \\\n",
       " 0  10X Dual Index TS Set A [PN-1000251]  10X Dual Index TS Set A [PN-1000251]   \n",
       " \n",
       "   index_type_id  \n",
       " 0             1  }),\n",
       " 'flex_annotation': StepFile(args={}, metadata={'project_name': 'Effect of a treatment on mouse CDX syngenic model', 'workflow': 'library_annotation', 'workflow_type': 'pooled', 'project_id': None, 'seq_request_id': 1, 'user_id': 3, 'project_description': 'Compared transcriptomic of treated CDX vs control\\n', 'assay_type_id': 15, 'mux_type_id': None, 'nuclei_isolation': False, 'pool_contact_name': 'Emilie Crouchet', 'pool_contact_email': 'ecrouchet@unistra.fr', 'pool_contact_phone': '+33(0)368853707'}, tables={'sample_pooling_table':   sample_name                library_name      sample_pool mux_barcode  \\\n",
       " 0        3671  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC001   \n",
       " 1        3672  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC002   \n",
       " 2        3709  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC003   \n",
       " 3        3710  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC004   \n",
       " \n",
       "    mux_type_id  \n",
       " 0            3  \n",
       " 1            3  \n",
       " 2            3  \n",
       " 3            3  , 'library_table':                  library_name      sample_name          genome  genome_id  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC  Mouse (GRCm38)          2   \n",
       " \n",
       "                library_type  library_type_id             pool seq_depth  \n",
       " 0  10X Flex Gene Expression                1  CDX_Melanoma_EC      None  , 'sample_table':   sample_name sample_id\n",
       " 0        3671      None\n",
       " 1        3672      None\n",
       " 2        3709      None\n",
       " 3        3710      None, 'pool_table':          pool_name       pool_label pool_id  num_m_reads_requested\n",
       " 0  CDX_Melanoma_EC  CDX_Melanoma_EC    None                  400.0, 'barcode_table':                  library_name index_well   kit_i7   name_i7 sequence_i7  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX         A6  1000251  SI-TS-A6  AAAGTTCTGA   \n",
       " \n",
       "     kit_i5   name_i5 sequence_i5  kit_i7_id  kit_i5_id  \\\n",
       " 0  1000251  SI-TS-A6  CGCAGTCCAC         22         22   \n",
       " \n",
       "                             kit_i7_name                           kit_i5_name  \\\n",
       " 0  10X Dual Index TS Set A [PN-1000251]  10X Dual Index TS Set A [PN-1000251]   \n",
       " \n",
       "   index_type_id  \n",
       " 0             1  , 'flex_table':        sample_name demux_name barcode_id\n",
       " 0  CDX_Melanoma_EC       3671      BC001\n",
       " 1  CDX_Melanoma_EC       3672      BC002\n",
       " 2  CDX_Melanoma_EC       3709      BC003\n",
       " 3  CDX_Melanoma_EC       3710      BC004}),\n",
       " 'sample_attribute_annotation': StepFile(args={}, metadata={'project_name': 'Effect of a treatment on mouse CDX syngenic model', 'workflow': 'library_annotation', 'workflow_type': 'pooled', 'project_id': None, 'seq_request_id': 1, 'user_id': 3, 'project_description': 'Compared transcriptomic of treated CDX vs control\\n', 'assay_type_id': 15, 'mux_type_id': None, 'nuclei_isolation': False, 'pool_contact_name': 'Emilie Crouchet', 'pool_contact_email': 'ecrouchet@unistra.fr', 'pool_contact_phone': '+33(0)368853707'}, tables={'sample_pooling_table':   sample_name                library_name      sample_pool mux_barcode  \\\n",
       " 0        3671  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC001   \n",
       " 1        3672  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC002   \n",
       " 2        3709  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC003   \n",
       " 3        3710  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC004   \n",
       " \n",
       "    mux_type_id  \n",
       " 0            3  \n",
       " 1            3  \n",
       " 2            3  \n",
       " 3            3  , 'library_table':                  library_name      sample_name          genome  genome_id  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC  Mouse (GRCm38)          2   \n",
       " \n",
       "                library_type  library_type_id             pool seq_depth  \\\n",
       " 0  10X Flex Gene Expression                1  CDX_Melanoma_EC      None   \n",
       " \n",
       "    is_mux_sample  is_flex_sample  \n",
       " 0          False            True  , 'sample_table':   sample_name sample_id _attr_tissue\n",
       " 0        3671      None         Skin\n",
       " 1        3672      None         Skin\n",
       " 2        3709      None         Skin\n",
       " 3        3710      None         Skin, 'pool_table':          pool_name       pool_label pool_id  num_m_reads_requested\n",
       " 0  CDX_Melanoma_EC  CDX_Melanoma_EC    None                  400.0, 'barcode_table':                  library_name index_well   kit_i7   name_i7 sequence_i7  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX         A6  1000251  SI-TS-A6  AAAGTTCTGA   \n",
       " \n",
       "     kit_i5   name_i5 sequence_i5  kit_i7_id  kit_i5_id  \\\n",
       " 0  1000251  SI-TS-A6  CGCAGTCCAC         22         22   \n",
       " \n",
       "                             kit_i7_name                           kit_i5_name  \\\n",
       " 0  10X Dual Index TS Set A [PN-1000251]  10X Dual Index TS Set A [PN-1000251]   \n",
       " \n",
       "   index_type_id  \n",
       " 0             1  , 'flex_table':        sample_name demux_name barcode_id\n",
       " 0  CDX_Melanoma_EC       3671      BC001\n",
       " 1  CDX_Melanoma_EC       3672      BC002\n",
       " 2  CDX_Melanoma_EC       3709      BC003\n",
       " 3  CDX_Melanoma_EC       3710      BC004}),\n",
       " 'complete_sas': StepFile(args={}, metadata={'project_name': 'Effect of a treatment on mouse CDX syngenic model', 'workflow': 'library_annotation', 'workflow_type': 'pooled', 'project_id': None, 'seq_request_id': 1, 'user_id': 3, 'project_description': 'Compared transcriptomic of treated CDX vs control\\n', 'assay_type_id': 15, 'mux_type_id': None, 'nuclei_isolation': False, 'pool_contact_name': 'Emilie Crouchet', 'pool_contact_email': 'ecrouchet@unistra.fr', 'pool_contact_phone': '+33(0)368853707'}, tables={'sample_pooling_table':   sample_name                library_name      sample_pool mux_barcode  \\\n",
       " 0        3671  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC001   \n",
       " 1        3672  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC002   \n",
       " 2        3709  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC003   \n",
       " 3        3710  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC004   \n",
       " \n",
       "    mux_type_id  \n",
       " 0            3  \n",
       " 1            3  \n",
       " 2            3  \n",
       " 3            3  , 'library_table':                  library_name      sample_name          genome  genome_id  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC  Mouse (GRCm38)          2   \n",
       " \n",
       "                library_type  library_type_id             pool seq_depth  \\\n",
       " 0  10X Flex Gene Expression                1  CDX_Melanoma_EC      None   \n",
       " \n",
       "    is_mux_sample  is_flex_sample  library_id  \n",
       " 0          False            True           1  , 'sample_table':   sample_name  sample_id _attr_tissue\n",
       " 0        3671          1         Skin\n",
       " 1        3672          2         Skin\n",
       " 2        3709          3         Skin\n",
       " 3        3710          4         Skin, 'pool_table':          pool_name       pool_label pool_id  num_m_reads_requested\n",
       " 0  CDX_Melanoma_EC  CDX_Melanoma_EC    None                  400.0, 'barcode_table':                  library_name index_well   kit_i7   name_i7 sequence_i7  \\\n",
       " 0  CDX_Melanoma_EC_10XFLEXGEX         A6  1000251  SI-TS-A6  AAAGTTCTGA   \n",
       " \n",
       "     kit_i5   name_i5 sequence_i5  kit_i7_id  kit_i5_id  \\\n",
       " 0  1000251  SI-TS-A6  CGCAGTCCAC         22         22   \n",
       " \n",
       "                             kit_i7_name                           kit_i5_name  \\\n",
       " 0  10X Dual Index TS Set A [PN-1000251]  10X Dual Index TS Set A [PN-1000251]   \n",
       " \n",
       "   index_type_id  \n",
       " 0             1  , 'flex_table':        sample_name demux_name barcode_id\n",
       " 0  CDX_Melanoma_EC       3671      BC001\n",
       " 1  CDX_Melanoma_EC       3672      BC002\n",
       " 2  CDX_Melanoma_EC       3709      BC003\n",
       " 3  CDX_Melanoma_EC       3710      BC004})}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "msf = pickle.load(open(\"/Users/agynter/Downloads/fb712880-c755-432d-8bcc-ec87ef4496a3.msf\", \"rb\"))\n",
    "msf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['header', 'project_select', 'pooled_library_annotation', 'pool_mapping', 'barcode_input', 'index_kit_mapping', 'flex_annotation', 'sample_attribute_annotation', 'complete_sas'])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "msf.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>sample_pool</th>\n",
       "      <th>mux_barcode</th>\n",
       "      <th>mux_type_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3671</td>\n",
       "      <td>CDX_Melanoma_EC_10XFLEXGEX</td>\n",
       "      <td>CDX_Melanoma_EC</td>\n",
       "      <td>BC001</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3672</td>\n",
       "      <td>CDX_Melanoma_EC_10XFLEXGEX</td>\n",
       "      <td>CDX_Melanoma_EC</td>\n",
       "      <td>BC002</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3709</td>\n",
       "      <td>CDX_Melanoma_EC_10XFLEXGEX</td>\n",
       "      <td>CDX_Melanoma_EC</td>\n",
       "      <td>BC003</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3710</td>\n",
       "      <td>CDX_Melanoma_EC_10XFLEXGEX</td>\n",
       "      <td>CDX_Melanoma_EC</td>\n",
       "      <td>BC004</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name                library_name      sample_pool mux_barcode  \\\n",
       "0        3671  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC001   \n",
       "1        3672  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC002   \n",
       "2        3709  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC003   \n",
       "3        3710  CDX_Melanoma_EC_10XFLEXGEX  CDX_Melanoma_EC       BC004   \n",
       "\n",
       "   mux_type_id  \n",
       "0            3  \n",
       "1            3  \n",
       "2            3  \n",
       "3            3  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "msf[\"complete_sas\"].tables[\"sample_pooling_table\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "opengsync",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
